File: RailgunVB/Core/Api/Youtube/YoutubeSearch.vb
Imports Google.Apis.Services
Imports Google.Apis.YouTube.v3
Imports Google.Apis.YouTube.v3.Data
Imports RailgunVB.Core.Configuration

Namespace Core.Api.Youtube
    
    Public Class YoutubeSearch
        
        Private ReadOnly _service As YouTubeService
        
        Public Sub New(config As MasterConfig)
            _service = New YouTubeService(New BaseClientService.Initializer() With {
                                             .ApiKey = config.GoogleApiToken,
                                             .ApplicationName = Me.GetType().ToString()
                                             })
        End Sub
    
        Public Async Function GetVideoAsync(query As String) As Task(Of YoutubeVideoData)
            Dim request As SearchResource.ListRequest = _service.Search.List("snippet")
            
            request.Q= query
            request.MaxResults = 1
            
            Dim response As SearchListResponse = Await request.ExecuteAsync()
            
            If response.Items.Count < 1 Then Return Nothing
            
            Dim result As SearchResult = response.Items.FirstOrDefault()
            
            If result Is Nothing OrElse Not (result.Id.Kind = "youtube#video") Then Return Nothing
            
            Dim snippet As SearchResultSnippet = result.Snippet
            
            Return New YoutubeVideoData(snippet.Title, result.Id.VideoId, snippet.ChannelTitle)
        End Function
        
    End Class
    
End NameSpace

File: Railgun/Apis/Youtube/YoutubeSearch.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Google.Apis.Services;
using Google.Apis.YouTube.v3;
using Railgun.Core.Configuration;

namespace Railgun.Apis.Youtube
{
    public class YoutubeSearch
    {
        private readonly YouTubeService _service;

        public YoutubeSearch(MasterConfig config) => _service = new YouTubeService(new BaseClientService.Initializer()
        {
            ApiKey = config.GoogleApiToken,
            ApplicationName = this.GetType().ToString()
        });

        public async Task<YoutubeVideoData> GetVideoAsync(string query)
        {
            var request = _service.Search.List("snippet");

            request.Q = query;
            request.MaxResults = 1;

            var response = await request.ExecuteAsync();

            if (response.Items.Count < 1) return null;

            var result = response.Items.FirstOrDefault();

            if (result == null || !(result.Id.Kind == "youtube#video")) return null;

            var snippet = result.Snippet;

            return new YoutubeVideoData(snippet.Title, result.Id.VideoId, snippet.ChannelTitle);
        }

        public async Task<IEnumerable<YoutubeVideoData>> GetVideosAsync(string playlistId)
        {
            var list = new List<YoutubeVideoData>();

            var nextPageToken = "";
            while (nextPageToken != null)
            {
                var request = _service.PlaylistItems.List("snippet");
                request.Id = playlistId;
                request.MaxResults = 50;
                request.PageToken = nextPageToken;

                var response = await request.ExecuteAsync();

                if (response.Items.Count < 1) return null;

                foreach (var item in response.Items)
                    list.Add(new YoutubeVideoData(item.Snippet.Title, item.ContentDetails.VideoId, item.Snippet.ChannelTitle));

                nextPageToken = response.NextPageToken;
            }

            return list;
        }
    }
}